# Working with data at scale: pandas

## From six paragraphs to seventeen hundred books

In the last notebook you built a search engine by hand. It worked — over **six** documents.

Everything you did there still applies. What changes today is only the **scale**, and therefore the tools. `pandas` is the library that lets you say what you want done to a whole column at once.

Today's corpus is: **1,750 books from the British Library's 19th-century collection**, each one labelled by hand as *Fiction* or *Non-fiction* by the Library's crowdsourced annotation project.

## Pandas to the rescue!

![pandas](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExaHNrd245NXpvMWp4NDVyaW15azNrZmxtemoyanlwbTNmNm1paTA1bSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/aCa8jFalHHJvi/giphy.gif)

(The name comes from "**pan**el **da**ta", a term from econometrics — not, sadly, from the animal. The animal has been adopted regardless.)

## Loading the data

On Colab we first download the file:

In [ ]:
!wget -q https://raw.githubusercontent.com/Living-with-machines/dhoxss-text2tech/main/Sessions/data/bl_books_genre.csv

### Enjoy this while the data downloads

![downloading](https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExYXdwbnBiaWMzZHQ3MTllc2h2N3YweWsybnd5eDUyNmU1NWwxYmt1eCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/BdMRkwomqVsLm/giphy.gif)

Then we import the library. `pandas` is conventionally imported as `pd` — you will see this abbreviation in every piece of pandas code you ever read:

In [ ]:
import pandas as pd

df = pd.read_csv("bl_books_genre.csv")

`df` is short for **DataFrame**, the pandas name for a table. Let's look at it.

`.head()` shows the first few rows:

In [2]:
df.head()

,record_id,title,author,date,place,country,language,publisher,genre,n_annotations
0,14603046,The Canadian farmer. A missionary incident [Si...,"Yates, William Joseph H.",1879.0,London,England,English,NaN,Fiction,5
1,14605595,"A new musical Interlude, called the Election [...",NaN,1774.0,London,England,English,NaN,Can't tell,5
2,14605643,An Elegy written among the ruins of an Abbey. ...,NaN,1765.0,London,England,English,J. Dodsley,Fiction,5
3,14616539,The Baron's Daughter. A ballad by the author o...,NaN,1877.0,Edinburgh,Scotland,English,"Ballantyne, Hanson",Fiction,5
4,14616550,"A Little Book of Verse, etc","Bell, Raimonde, pseudonym [i.e. James Pickerin...",1893.0,London,England,English,Poetry & Prose,Fiction,4


`.shape` gives the number of rows and columns, and `.columns` lists the column names:

In [3]:
print(df.shape)
print(df.columns)

(1750, 10)
Index(['record_id', 'title', 'author', 'date', 'place', 'country', 'language',
       'publisher', 'genre', 'n_annotations'],
      dtype='object')


So: 1,750 rows (one per book) and 10 columns (one per thing we know about each book).

⚠️ Notice how much this resembles a **dictionary of lists**, which is more or less what a DataFrame is underneath. The difference is everything pandas adds on top.

## One column at a time

A single column is selected with square brackets and its name:

In [4]:
df["title"]

0       The Canadian farmer. A missionary incident [Si...
1       A new musical Interlude, called the Election [...
2       An Elegy written among the ruins of an Abbey. ...
3       The Baron's Daughter. A ballad by the author o...
4                             A Little Book of Verse, etc
                              ...                        
1745    Forhandlingsprotokol ført i Regjeringsraadet o...
1746    De første Konger af den Oldenborgske Slægt-Omr...
1747    Select Statutes and other Constitutional Docum...
1748            The Poetical Works of Robert Southey, etc
1749    Anteckningar rörande Expeditionen åt Westerbot...
Name: title, Length: 1750, dtype: object

A column is called a **Series**. It behaves like a list you can do arithmetic on — and, crucially, like something you can *count*.

`.value_counts()` is the pandas version of the `Counter` from notebook 2e:

In [5]:
df["genre"].value_counts()

genre
Non-fiction    1305
Fiction         439
Both              4
Can't tell        2
Name: count, dtype: int64

## Sequencing

`value_counts()` is a method or function you apply to a column or `pd.Series'. You can apply multiple methods in **sequence**.

This shows how with little code you can quickly perform interesting analysis with pandas.

In [8]:
df["genre"].value_counts(normalize=True).plot(kind='bar')

ImportError: matplotlib is required for plotting when the default backend "matplotlib" is selected.

Straight away, something important: the collection is **heavily unbalanced**. There are roughly three times as many non-fiction books as fiction. Remember this — in the next notebook it will nearly fool us.

Let's look at the collection in other ways:

In [9]:
print(df["language"].value_counts().head(6))

language
English    845
Dutch      332
Swedish    189
Danish     165
German     102
Spanish     21
Name: count, dtype: int64


In [10]:
print(df["place"].value_counts().head(6))

place
London        543
Stockholm     121
Kjøbenhavn    100
Amsterdam      77
New York       40
Haarlem        31
Name: count, dtype: int64


This is worth pausing on. This is a *British Library* collection, but fewer than half its books are in English: there is a large Dutch, Swedish, Danish and German presence, published in Amsterdam, Stockholm and København.

A collection is never a neutral sample of the past. It is the result of what somebody chose to buy, keep and digitise — and that will show up in every number we compute from it today.

### ✏️ Exercise 1

Print the ten most common **publishers** in the collection, and the number of books published in each **country**.

In [ ]:
# Type your code here:


## Selecting data

Similar to lists and dictionaries, Pandas comes with many inbuilt tools to select data from a dataframe.

We won't go into too much detail here, but the main tools here are `.iloc` and `.loc`

In [13]:
df.iloc[0]

record_id                                                 14603046
title            The Canadian farmer. A missionary incident [Si...
author                                    Yates, William Joseph H.
date                                                        1879.0
place                                                       London
country                                                    England
language                                                   English
publisher                                                      NaN
genre                                                      Fiction
n_annotations                                                    5
Name: 0, dtype: object

A common error when selecting multiple columns is forgetting the extra square bracket.

In [20]:
df['title','language'] # this doesn't work

KeyError: ('title', 'language')

In [21]:
df[['title','language']] # this works

,title,language
0,The Canadian farmer. A missionary incident [Si...,English
1,"A new musical Interlude, called the Election [...",English
2,An Elegy written among the ruins of an Abbey. ...,English
3,The Baron's Daughter. A ballad by the author o...,English
4,"A Little Book of Verse, etc",English
...,...,...
1745,Forhandlingsprotokol ført i Regjeringsraadet o...,Danish
1746,De første Konger af den Oldenborgske Slægt-Omr...,Danish
1747,Select Statutes and other Constitutional Docum...,English
1748,"The Poetical Works of Robert Southey, etc",English


You can combine row and column names to retrieve a specific segemnt of the 

In [17]:
df.loc[1:4,['title','language']]

,title,language
1,"A new musical Interlude, called the Election [...",English
2,An Elegy written among the ruins of an Abbey. ...,English
3,The Baron's Daughter. A ballad by the author o...,English
4,"A Little Book of Verse, etc",English


### ✏️ Exercise 2

Can you retrieve the last two titles and their publishers?

In [ ]:
# Type your code here:

1748            The Poetical Works of Robert Southey, etc
1749    Anteckningar rörande Expeditionen åt Westerbot...
Name: title, dtype: object

1745    Forhandlingsprotokol ført i Regjeringsraadet o...
1746    De første Konger af den Oldenborgske Slægt-Omr...
1747    Select Statutes and other Constitutional Docum...
1748            The Poetical Works of Robert Southey, etc
1749    Anteckningar rörande Expeditionen åt Westerbot...
Name: title, dtype: object

## Selecting rows: masks

To select *rows*, we write a condition. The condition produces a column of `True`/`False` or boolean values.

This is called a **mask** — and pandas keeps the rows evaluate to `True`.

In [ ]:
# Which rows are in English?
mask = df["language"] == "English"

print(mask.head())


0    True
1    True
2    True
3    True
4    True
Name: language, dtype: bool

English books: 845


You can perform operation on the **mask**, for example to count how many English language books we have in the dataframe.

`True` counts as 1, so summing counts the matches

In [ ]:

print("English books:", mask.sum()) 

Putting the mask in square brackets gives us those rows as a new DataFrame:

In [ ]:
english = df[df["language"] == "English"]

print(english.shape)
english.head(3)

Conditions can be combined with `&` (and) and `|` (or). Each condition needs its own parentheses:

In [ ]:
early = df[(df["date"] < 1800) & (df["language"] == "English")]

print("English books published before 1800:", len(early))
early[["title", "date", "place"]].head()

☝️ Note `df[["title", "date", "place"]]`: a *list* of column names inside the brackets selects several columns at once.

### ✏️ Exercise 3

How many books in the collection are **fiction published in London**? Print the number, and the titles of the first five.

In [ ]:
# Type your code here:


## Sorting

`.sort_values()` orders the table by a column — the pandas version of the `sorted()` you used for ranking search results:

In [ ]:
df.sort_values("date").head(3)[["title", "date"]]

In [ ]:
# ... and the most recent:
df.sort_values("date", ascending=False).head(3)[["title", "date"]]

## Working with text in a column

Here is where the week comes back together. Every string method you learnt in notebook 2a — `.lower()`, `.split()`, `.count()`, `.startswith()` — exists in pandas as `.str.something`, and applies to **the whole column at once**.

In [23]:
# How many characters in each title?
df["title"].str.len().head()

0     90
1     64
2     86
3    100
4     27
Name: title, dtype: int64

In [24]:
# How many words in each title? (.split() then count the pieces)
df["title"].str.split().str.len().head()

0    16
1    11
2    17
3    17
4     6
Name: title, dtype: int64

We can store the result as a **new column**, simply by assigning to a name that does not exist yet:

In [27]:
df["title_words"] = df["title"].str.split().str.len()

df[["title", "title_words"]].head()

,title,title_words
0,The Canadian farmer. A missionary incident [Si...,16
1,"A new musical Interlude, called the Election [...",11
2,An Elegy written among the ruins of an Abbey. ...,17
3,The Baron's Daughter. A ballad by the author o...,17
4,"A Little Book of Verse, etc",6


`.str.contains()` searches inside every title at once. It returns a mask, so we can use it to select rows:

In [28]:
has_history = df["title"].str.contains("history", case=False)

print("titles containing 'history':", has_history.sum())
df[has_history][["title", "genre"]].head()

titles containing 'history': 99


,title,genre
50,"Tamworth Castle. Its foundation, its history, ...",Non-fiction
52,The History of the Royal Pavilion at Brighton,Non-fiction
55,The Early History of Spennymoor. A paper read ...,Non-fiction
76,A Short History of Clent,Non-fiction
104,The Dawn of Modern Geography. A history of exp...,Non-fiction


### ✏️ Exercise 4

In notebook 2g we searched a corpus of six documents for a word. Do it again, over 1,750 titles:

1. How many titles contain the word `"tale"`? (Careful with upper and lower case.)
2. Of those, how many are labelled *Fiction*?
3. Try the same with `"history"`. What do you notice?

In [ ]:
# Type your code here:


### `.apply()`: your own functions, over a whole column

`.str` covers the common operations. When you need something it does not offer, you write a **function** — exactly as in notebook 1d — and hand it to `.apply()`, which runs it on every row of the column.

In [29]:
def first_word(title):
    """ return the first word of a title, lowercased, without punctuation

    Args:
        title: the title of a book, as a string

    Returns:
        The first word, as a string
    """
    words = title.lower().split()
    return words[0].strip(".,;:'\"")


# Try it on a single title first — always do this before applying it to 1,750 of them:
print(first_word("The Canadian farmer. A missionary incident"))

the


Now apply it to the whole column:

In [30]:
df["first_word"] = df["title"].apply(first_word)

df["first_word"].value_counts().head(10)

first_word
the             209
a               111
de               48
geschiedenis     29
history          21
het              17
geschichte       16
bidrag           16
om               15
an               13
Name: count, dtype: int64

⚠️ `.apply()` is the general escape hatch: anything you can write as a function, you can run over a whole column. It is slower than the built-in `.str` methods, so prefer those when they exist — but `.apply()` is what makes pandas open-ended rather than a fixed menu.

### ✏️ Exercise 5

Write a function called `is_long_title` that takes a title and returns `True` if it has more than 15 words, `False` otherwise. Apply it to the `title` column, store the result as a new column, and count how many long titles there are.

In [ ]:
# Type your code here:


### ✏️ Exercise 6

Write a function that computes the average word length, and apply it to the "title" `column`.

In [ ]:
# Type your code here:

## Grouping: asking questions of subsets

`.groupby()` splits the table into groups, computes something for each, and puts the answers back together. It is the single most useful thing pandas does — and the picture explains it better than the words:

![split-apply-combine](https://jakevdp.github.io/PythonDataScienceHandbook/figures/03.08-split-apply-combine.png)

*(Figure from Jake VanderPlas, [Python Data Science Handbook](https://jakevdp.github.io/PythonDataScienceHandbook/).)*

**Split** the rows into groups, **apply** a calculation to each group, **combine** the answers into one table.

For example: do fiction and non-fiction titles differ in length?

In [31]:
df.groupby("genre")["title_words"].mean()

genre
Both            9.500000
Can't tell     13.000000
Fiction         9.214123
Non-fiction    14.003831
Name: title_words, dtype: float64

They do — and in a direction worth thinking about. Now: how is genre distributed across languages?

In [32]:
# `.isin()` keeps rows whose value is in a list — handy when a column has a long tail:
main_languages = ["English", "Dutch", "Swedish", "Danish", "German"]

df[df["language"].isin(main_languages)].groupby("language")["genre"].value_counts()

language  genre      
Danish    Non-fiction    164
          Fiction          1
Dutch     Non-fiction    254
          Fiction         78
English   Non-fiction    518
          Fiction        322
          Both             4
          Can't tell       1
German    Non-fiction     80
          Fiction         22
Swedish   Non-fiction    188
          Can't tell       1
Name: count, dtype: int64

Look at that table closely, because it matters more than it looks.

The Danish books are 164 non-fiction to **1** fiction. The Swedish ones are 188 non-fiction to **none at all**. The English part of the collection is a mix; the non-English part is almost entirely non-fiction.

So in this collection, *language* and *genre* are tangled together. Any conclusion we draw about genre might really be a conclusion about language. Keep this in your pocket — in the next notebook it comes back to bite us.

### Trends over time

To look at change over time we need decades rather than individual years. Integer division (`//`) rounds down, so `1857 // 10 * 10` gives `1850`:

In [33]:
df["decade"] = (df["date"] // 10) * 10

books_per_decade = df.groupby("decade").size()
print(books_per_decade.tail(10))

decade
1830.0     72
1840.0    116
1850.0    197
1860.0    187
1870.0    232
1880.0    303
1890.0    512
1900.0     12
1910.0      1
1970.0      1
dtype: int64


In [34]:
# pandas can plot directly:
books_per_decade.plot(kind="bar", figsize=(12, 4), title="Books in the collection, by decade")

ImportError: matplotlib is required for plotting when the default backend "matplotlib" is selected.

⚠️ Read that chart carefully. It does **not** show that more books were written in the 1880s. It shows that more books *from* the 1880s are in **this collection** — a fact about the British Library's digitisation choices, not about the nineteenth century.

### ✏️ Exercise 7

Using `.groupby()`, find out **how the proportion of fiction changes over time**.

Suggested steps:
* create a column `is_fiction` that is `True` when `genre` is `"Fiction"`
* group by `decade` and take the `.mean()` of that column (the mean of a True/False column is the proportion of `True`)
* plot it

Then ask yourself the same question as above: what does the result actually describe?

In [ ]:
# Type your code here:


## What's next

You now have a table where every row is a book and every column is something we know — or something we measured. We asked it questions with masks, `.str` methods, our own functions and `.groupby()`.

In the next notebook we turn the question around. Instead of *asking* how many fiction books there are, we hide the `genre` column and ask a computer to **work it out from the title alone**.

# Solutions

### ✏️ Exercise 1

In [ ]:
print(df["publisher"].value_counts().head(10))
print()
print(df["country"].value_counts())

### ✏️ Exercise 2

In [ ]:
display(df.loc[len(df)-2:len(df)-1, 'title'])
display(df.tail()['title'])

### ✏️ Exercise 3

In [ ]:
london_fiction = df[(df["genre"] == "Fiction") & (df["place"] == "London")]

print("Fiction published in London:", len(london_fiction))
print(london_fiction["title"].head().to_list())

### ✏️ Exercise 4

In [ ]:
tale = df["title"].str.contains("tale", case=False)
print("titles containing 'tale':", tale.sum())
print(df[tale]["genre"].value_counts())
print()

history = df["title"].str.contains("history", case=False)
print("titles containing 'history':", history.sum())
print(df[history]["genre"].value_counts())

# 'tale' points strongly to fiction, 'history' strongly to non-fiction. Single words
# carry real information about genre — which is exactly what the next notebook exploits.
# Note also that 'tale' matches 'tales' and 'Tale', but it would also match a word like
# 'talented' if one appeared: `.str.contains()` looks for the letters, not the word.

### ✏️ Exercise 5

In [ ]:
def is_long_title(title):
    """ True if the title has more than 15 words """
    return len(title.split()) > 15


df["long_title"] = df["title"].apply(is_long_title)
print(df["long_title"].value_counts())

### ✏️ Exercise 6

In [42]:
def averge_word_length(text):
    words = text.split()
    word_lengths = [len(w) for w in words]
    return sum(word_lengths) / len(word_lengths)
    
df['av_word_length'] =df['title'].apply(averge_word_length)
df[['title','av_word_length']].sort_values(by='av_word_length', ascending=False)

,title,av_word_length
1442,Minnes-bilder,13.0
901,Gamle Christiania-Billeder,12.5
754,Italië. Reisherinneringen,12.0
1299,Geschiedkundige opstellen,12.0
1617,Geschiedenis,12.0
...,...,...
590,The Quest of a Heart,3.2
454,Happy with Either. A novel. By A. L. O. S,3.2
420,At Heart a Rake,3.0
160,Too Apt a Pupil,3.0


### ✏️ Exercise 7

In [ ]:
df["is_fiction"] = df["genre"] == "Fiction"

fiction_by_decade = df.groupby("decade")["is_fiction"].mean()
print(fiction_by_decade.tail(10))

fiction_by_decade.plot(figsize=(12, 4), title="Proportion of fiction, by decade")

# Again: this is the proportion of fiction *in this collection*, decade by decade.
# Decades with only a handful of books will bounce around wildly — a proportion
# computed from three books is not a trend.